api.go

```diff
@@ -0,0 +1,258 @@
+package nsone
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+)
+
+const (
+	clientVersion    = "0.9.0"
+	defaultEndpoint  = "https://api.nsone.net/v1/"
+	defaultUserAgent = "go-ns1/" + clientVersion
+
+	headerAuth          = "X-NSONE-Key"
+	headerRateLimit     = "X-Ratelimit-Limit"
+	headerRateRemaining = "X-Ratelimit-Remaining"
+	headerRatePeriod    = "X-Ratelimit-Period"
+)
+
+// Doer is a single method interface that allows a user to extend/augment an http.Client instance.
+// Note: http.Client satisfies the Doer interface.
+type Doer interface {
+	Do(*http.Request) (*http.Response, error)
+}
+
+// APIClient stores NS1 client state
+type APIClient struct {
+	// client handles all http communication.
+	// The default value is *http.Client
+	client Doer
+
+	// NS1 rest endpoint, overrides default if given.
+	Endpoint *url.URL
+
+	// NS1 api key (value for http request header 'X-NSONE-Key').
+	ApiKey string
+
+	// NS1 go rest user agent (value for http request header 'User-Agent').
+	UserAgent string
+
+	// Func to call after response is returned in Do
+	RateLimitFunc func(RateLimit)
+
+	// Enables verbose logs.
+	debug bool
+}
+
+// New takes an API Key and creates an *APIClient
+func New(k string) *APIClient {
+	endpoint, _ := url.Parse(defaultEndpoint)
+	return &APIClient{
+		client:        http.DefaultClient,
+		Endpoint:      endpoint,
+		ApiKey:        k,
+		RateLimitFunc: defaultRateLimitFunc,
+		UserAgent:     defaultUserAgent,
+	}
+}
+
+func NewAPIClient(httpClient Doer, options ...APIClientOption) *APIClient {
+	endpoint, _ := url.Parse(defaultEndpoint)
+
+	if httpClient == nil {
+		httpClient = http.DefaultClient
+	}
+
+	c := &APIClient{
+		client:        httpClient,
+		Endpoint:      endpoint,
+		RateLimitFunc: defaultRateLimitFunc,
+		UserAgent:     defaultUserAgent,
+	}
+
+	for _, option := range options {
+		option(c)
+	}
+	return c
+}
+
+// Debug enables debug logging
+func (c *APIClient) Debug() {
+	c.debug = true
+}
+
+type APIClientOption func(*APIClient)
```

Comment:
I wasnt sure if this was the best/right way to handle optional parameters(by passing in closures that set the attribute on the client), but I liked it after reading these:

http://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis
https://github.com/digitalocean/godo/issues/84